src/uct/cuda/gdr_copy/gdr_copy_md.c

```diff
@@ -30,77 +61,377 @@ static ucs_status_t uct_gdr_copy_md_query(uct_md_h md, uct_md_attr_t *md_attr)
 static ucs_status_t uct_gdr_copy_mkey_pack(uct_md_h md, uct_mem_h memh,
                                            void *rkey_buffer)
 {
+    uct_gdr_copy_key_t *packed      = (uct_gdr_copy_key_t *)rkey_buffer;
+    uct_gdr_copy_mem_t *mem_hndl    = (uct_gdr_copy_mem_t *)memh;
+
+    packed->vaddr   = mem_hndl->info.va;
+    packed->bar_ptr = mem_hndl->bar_ptr;
+
     return UCS_OK;
 }
 
 static ucs_status_t uct_gdr_copy_rkey_unpack(uct_md_component_t *mdc,
                                              const void *rkey_buffer, uct_rkey_t *rkey_p,
                                              void **handle_p)
 {
-    *rkey_p   = 0xdeadbeef;
+    uct_gdr_copy_key_t *packed = (uct_gdr_copy_key_t *)rkey_buffer;
+    uct_gdr_copy_key_t *key;
+
+    key = ucs_malloc(sizeof(uct_gdr_copy_key_t), "uct_gdr_copy_key_t");
+    if (NULL == key) {
+        ucs_error("Failed to allocate memory for uct_gdr_copy_key_t");
+        return UCS_ERR_NO_MEMORY;
+    }
+    key->vaddr      = packed->vaddr;
+    key->bar_ptr    = packed->bar_ptr;
+
     *handle_p = NULL;
+    *rkey_p   = (uintptr_t)key;
+
     return UCS_OK;
 }
 
 static ucs_status_t uct_gdr_copy_rkey_release(uct_md_component_t *mdc, uct_rkey_t rkey,
                                               void *handle)
 {
+    ucs_assert(NULL == handle);
+    ucs_free((void *)rkey);
     return UCS_OK;
 }
 
-static ucs_status_t uct_gdr_copy_mem_reg(uct_md_h md, void *address, size_t length,
+static ucs_status_t
+uct_gdr_copy_mem_reg_internal(uct_md_h uct_md, void *address, size_t length,
+                              unsigned flags, uct_gdr_copy_mem_t *mem_hndl)
+{
+    uct_gdr_copy_md_t *md = ucs_derived_of(uct_md, uct_gdr_copy_md_t);
+    CUdeviceptr d_ptr = ((CUdeviceptr )(char *) address);
+    gdr_mh_t mh;
+    void *bar_ptr;
+    gdr_info_t info;
+
+    if (!length) {
+        mem_hndl->mh = 0;
+        return UCS_OK;
+    }
+
+    if (gdr_pin_buffer(md->gdrcpy_ctx, d_ptr, length, 0, 0, &mh) != 0) {
+        ucs_error("gdr_pin_buffer Failed. length :%lu ", length);
+        return UCS_ERR_IO_ERROR;
+    }
+    if (mh == 0) {
+        ucs_error("gdr_pin_buffer Failed. length :%lu ", length);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if (gdr_map(md->gdrcpy_ctx, mh, &bar_ptr, length) !=0) {
+        ucs_error("gdr_map failed. length :%lu ", length);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if (gdr_get_info(md->gdrcpy_ctx, mh, &info) != 0) {
+        ucs_error("gdr_get_info failed. ");
```

Comment:
extra space in the end
pls print error codes in all error flows (multiple places)